python/cudf/udf_cpp/groupby/function.cu

```diff
@@ -0,0 +1,623 @@
+}
+
+// Calculate minimum of the group, return the scalar
+extern "C" __device__ int BlockMin_float64(double* numba_return_value,
+                                           double const* data,
+                                           int64_t size)
+{
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + blockDim.x - 1) / blockDim.x;
+
+  __shared__ double smin;
+
+  if (threadIdx.x == 0) { smin = DBL_MAX; }
+
+  __syncthreads();
+
+  device_min<double>(data, items_per_thread, size, DBL_MAX, &smin);
+
+  *numba_return_value = smin;
+
+  return 0;
+}
+
+// Calculate minimum of the group, return the scalar
+extern "C" __device__ int BlockIdxMax_int64(int64_t* numba_return_value,
+                                            int64_t const* data,
+                                            int64_t* index,
+                                            int64_t size)
+{
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + blockDim.x - 1) / blockDim.x;
+
+  __shared__ int64_t smax;
+  __shared__ int64_t sidx;
+
+  if (threadIdx.x == 0) {
+    smax = INT64_MIN;
+    sidx = INT64_MAX;
+  }
+
+  __syncthreads();
+
+  device_idxmax<int64_t>(data, items_per_thread, index, size, INT64_MIN, &smax, &sidx);
+
+  *numba_return_value = sidx;
+
+  return 0;
+}
+
+// Calculate minimum of the group, return the scalar
+extern "C" __device__ int BlockIdxMax_float64(int64_t* numba_return_value,
+                                              double const* data,
+                                              int64_t* index,
+                                              int64_t size)
+{
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + blockDim.x - 1) / blockDim.x;
+
+  __shared__ double smax;
+  __shared__ int64_t sidx;
+
+  if (threadIdx.x == 0) {
+    smax = -DBL_MAX;
+    sidx = INT64_MAX;
+  }
+
+  __syncthreads();
+
+  device_idxmax<double>(data, items_per_thread, index, size, -DBL_MAX, &smax, &sidx);
+
+  *numba_return_value = smax;
+
+  return 0;
+}
+
+// Calculate minimum of the group, return the scalar
+extern "C" __device__ int BlockIdxMin_int64(int64_t* numba_return_value,
+                                            int64_t const* data,
+                                            int64_t* index,
+                                            int64_t size)
+{
+  // Calculate how many elements each thread is working on
+  auto const items_per_thread = (size + blockDim.x - 1) / blockDim.x;
+
+  __shared__ int64_t smin;
+  __shared__ int64_t sidx;
+
+  if (threadIdx.x == 0) {
+    smin = INT64_MAX;
+    sidx = INT64_MAX;
+  }
+
+  __syncthreads();
+
+  device_idxmin<int64_t>(data, items_per_thread, index, size, INT64_MAX, &smin, &sidx);
+
+  *numba_return_value = sidx;
+
+  return 0;
+}
+
+// Calculate minimum of the group, return the scalar
+extern "C" __device__ int BlockIdxMin_float64(int64_t* numba_return_value,
```

Comment:
Would love to find a way to reduce all the copy-paste for these definitions.

Reply:
Done